### PR TITLE
MGMT-9825: create preflight package, responsible for preflight check …

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
…for single CR

This PR includes the following:
1) preflight package that executes preflight check for single PR
   a) compile helm chart
   b) get list of all yamls from the compiled chart
   c) go over the list of yamls and validate daemonset image, in case buildconfig is not present
2) add Init API to runtime, used by preflight
3) add mock file for helmer package

### CODE REVIEW

File: pkg/preflight/preflight.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+	metricsAPI     metrics.Metrics
+	runtimeAPI     runtime.RuntimeAPI
+	kernelAPI      kernel.KernelData
+}
+
+func (p *preflight) PreflightUpgradeCheck(ctx context.Context,
+	sr *srov1beta1.SpecialResource,
+	runInfo *runtime.RuntimeInformation) error {
+
+	p.log.Info("start preflight for cr", "name", sr.Name)
```
Comment (pmtk) [inline pkg/preflight/preflight.go:73]: nit: other logs start with uppercase letter

File: pkg/preflight/preflight.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+	k8sv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+//go:generate mockgen -source=preflight.go -package=preflight -destination=mock_preflight_api.go
+
+type PreflightAPI interface {
+	PreflightUpgradeCheck(ctx context.Context,
```
Comment (pmtk) [inline pkg/preflight/preflight.go:30]: I don't see it being used anywhere. This will be integrated in the future?

**Comment by ybettan** [comment]
Typo in the commit message:
preflight package that executes preflight check for single **PR** --> preflight package that executes preflight check for single **CR**

File: pkg/preflight/preflight.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+func (p *preflight) PreflightUpgradeCheck(ctx context.Context,
+	sr *srov1beta1.SpecialResource,
+	runInfo *runtime.RuntimeInformation) error {
+
+	p.log.Info("start preflight for cr", "name", sr.Name)
+	sr.DeepCopyInto(&runInfo.SpecialResource)
+
+	chart, err := p.helmerAPI.Load(sr.Spec.Chart)
+	if err != nil {
+		p.log.Info("Failed to load helm chart for CR", "name", sr.Name)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:78]:
```suggestion
		p.log.WithError(err).Error("Failed to load helm chart for CR", "name", sr.Name)
```

File: pkg/preflight/preflight.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	yamlsList, err := p.processFullChartTemplates(ctx, chart, sr.Spec.Set, runInfo, sr.Namespace, runInfo.KernelFullVersion)
+	if err != nil {
+		p.log.Info("Failed to process full chart during preflight check")
+		return err
+	}
+	verified, err := p.handleYAMLsCheck(ctx, yamlsList, runInfo.KernelFullVersion)
+	if err != nil {
+		p.log.Info("Failed to verify the chart during preflight check", "err", err)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:89]:
```suggestion
		p.log.Error(err, "Failed to verify the chart during preflight check", "err", err)
```

File: pkg/preflight/preflight.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+}
+
+func (p *preflight) processFullChartTemplates(ctx context.Context,
+	chart *helmchart.Chart,
+	values unstructured.Unstructured,
+	runInfo *runtime.RuntimeInformation,
+	namespace string,
+	upgradeKernelVersion string) (string, error) {
+	var err error
+	// [TODO] - check that templates are not messed up during helm processing
```
Comment (ybettan) [inline pkg/preflight/preflight.go:105]: Are we doing it in another PR?

File: pkg/preflight/preflight.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	nodeVersion := runInfo.ClusterUpgradeInfo[upgradeKernelVersion]
+
+	runInfo.ClusterVersionMajorMinor = nodeVersion.ClusterVersion
+	runInfo.OperatingSystemDecimal = nodeVersion.OSVersion
+	runInfo.OperatingSystemMajorMinor = nodeVersion.OSMajorMinor
+	runInfo.OperatingSystemMajor = nodeVersion.OSMajor
+	runInfo.DriverToolkitImage = nodeVersion.DriverToolkit.ImageURL
+
+	fullChart.Values, err = chartutil.CoalesceValues(&fullChart, values.Object)
+	if err != nil {
```
Comment (ybettan) [inline pkg/preflight/preflight.go:121]: Can we log the error?

File: pkg/preflight/preflight.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+		return "", err
+	}
+	rinfo, err := k8sruntime.DefaultUnstructuredConverter.ToUnstructured(&runInfo)
+	if err != nil {
```
Comment (ybettan) [inline pkg/preflight/preflight.go:125]: Log the error?

File: pkg/preflight/preflight.go [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+		return "", err
+	}
+
+	fullChart.Values, err = chartutil.CoalesceValues(&fullChart, rinfo)
+	if err != nil {
```
Comment (ybettan) [inline pkg/preflight/preflight.go:130]: Log the error?

File: pkg/preflight/preflight.go [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+		return "", err
+	}
+
+	return p.helmerAPI.GetHelmOutput(ctx, fullChart, fullChart.Values, namespace)
+}
+
+//[TODO] - handle multiple daemonsets and buildconfigs
```
Comment (ybettan) [inline pkg/preflight/preflight.go:137]: In a different PR?

File: pkg/preflight/preflight.go [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+func (p *preflight) handleYAMLsCheck(ctx context.Context, yamlsList string, upgradeKernelVersion string) (bool, error) {
+	objList, err := p.resourceAPI.GetObjectsFromYAML([]byte(yamlsList))
+	if err != nil {
```
Comment (ybettan) [inline pkg/preflight/preflight.go:140]: Log?

File: pkg/preflight/preflight.go [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+		p.log.Info("dtk info present in layer", "layerIndex", i)
+		if dtk.KernelFullVersion != upgradeKernelVersion {
+			p.log.Info("DTK kernel version differs from the upgrade node version", "dtkVersion", dtk.KernelFullVersion, "upgradeVersion", upgradeKernelVersion)
+			return false, nil
+		}
+		return true, nil
+	}
+
+	p.log.Info("DTK info not present on any layer of the image, not good", "image", image)
```
Comment (pmtk) [inline pkg/preflight/preflight.go:203]: This means that slim driver containers image (for example two stage builds with only *.ko and kmod in final image) are considered unverified but will still work?

File: pkg/preflight/preflight.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+	for _, obj := range objList.Items {
+		kind := obj.GetKind()
+		if kind == "BuildConfig" {
+			// no more need to check daemons set, build config is present
+			p.log.Info("preflight: buildconfig related to daemonset, skipping image verification")
+			break
+		}
+		if kind == "DaemonSet" {
+			verified, err = p.daemonSetPreflightCheck(ctx, &obj, upgradeKernelVersion)
+			if err != nil {
```
Comment (ybettan) [inline pkg/preflight/preflight.go:153]: Log?

File: pkg/preflight/preflight.go [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+		}
+	}
+	return verified, nil
+}
+
+func (p *preflight) daemonSetPreflightCheck(ctx context.Context, obj *unstructured.Unstructured, upgradeKernelVersion string) (bool, error) {
+	var ds k8sv1.DaemonSet
+	err := k8sruntime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &ds)
+	if err != nil {
+		p.log.Info("failed to convert YAML daemonset into sturct daemonset", "error", err)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:166]:
```suggestion
		p.log.Error(er, "failed to convert YAML daemonset into sturct daemonset", "error", err)
```

File: pkg/preflight/preflight.go [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+		return false, err
+	}
+	if ds.Spec.Template.Spec.Containers == nil || len(ds.Spec.Template.Spec.Containers) == 0 {
+		p.log.Info("invalid daemonset, no container  data present")
```
Comment (ybettan) [inline pkg/preflight/preflight.go:170]:
```suggestion
		p.log.Error(nil, "invalid daemonset, no container data present")
```

File: pkg/preflight/preflight.go [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+			return false, nil
+		}
+		return true, nil
+	}
+
+	p.log.Info("DTK info not present on any layer of the image, not good", "image", image)
+	return false, nil
+}
+
+func (p *preflight) PrepareRuntimeInfo(ctx context.Context, image string) (*runtime.RuntimeInformation, error) {
```
Comment (pmtk) [inline pkg/preflight/preflight.go:207]: Question: So this is a hypothetical runtime info after the upgrade?

File: pkg/preflight/preflight.go [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+		p.log.Info("invalid daemonset, no container  data present")
+		return false, fmt.Errorf("invalid daemonset, no container  data present")
+	}
+	image := ds.Spec.Template.Spec.Containers[0].Image
+
+	p.log.Info("daemonset image for preflight validation", "image", image)
+
+	repo, digests, auth, err := p.registryAPI.GetLayersDigests(ctx, image)
+	if err != nil {
+		p.log.Info("Failed to get layers digests for image", "image", image, "err", err)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:179]:
```suggestion
		p.log.Error(err, "Failed to get layers digests for image", "image", image)
```

File: pkg/preflight/preflight.go [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+		return false, nil
+	}
+
+	for i := len(digests) - 1; i >= 0; i-- {
+		layer, err := p.registryAPI.GetLayerByDigest(repo, digests[i], auth)
+		if err != nil {
+			p.log.Info("Failed to extract/access image", "image", image, "err", err)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:186]:
```suggestion
			p.log.Error(err, "Failed to extract/access image", "image", image)
```

File: pkg/preflight/preflight.go [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+			return false, nil
+		}
+		dtk, err := p.registryAPI.ExtractToolkitRelease(layer)
+		if err != nil {
+			p.log.Info("dtk info not present", "image", image, "layerIndex", i)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:191]:
```suggestion
			p.log.Error(err, "dtk info not present", "image", image, "layerIndex", i)
```

File: pkg/preflight/preflight.go [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+	p.log.Info("daemonset image for preflight validation", "image", image)
+
+	repo, digests, auth, err := p.registryAPI.GetLayersDigests(ctx, image)
+	if err != nil {
+		p.log.Info("Failed to get layers digests for image", "image", image, "err", err)
+		return false, nil
+	}
+
+	for i := len(digests) - 1; i >= 0; i-- {
```
Comment (ybettan) [inline pkg/preflight/preflight.go:183]: Not critical at all but is there a reason we go backward?

File: pkg/preflight/preflight.go [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		dtk, err := p.registryAPI.ExtractToolkitRelease(layer)
+		if err != nil {
+			p.log.Info("dtk info not present", "image", image, "layerIndex", i)
+			continue
+		}
+
+		p.log.Info("dtk info present in layer", "layerIndex", i)
+		if dtk.KernelFullVersion != upgradeKernelVersion {
+			p.log.Info("DTK kernel version differs from the upgrade node version", "dtkVersion", dtk.KernelFullVersion, "upgradeVersion", upgradeKernelVersion)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:197]: Not sure if we should log as an error here as well.

File: pkg/preflight/preflight.go [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+		return "", err
+	}
+
+	return dtk.KernelFullVersion, err
+}
+
+func (p *preflight) alertUserIfNeeded(verified bool, crName string) {
+	if verified {
+		p.log.Info("preflight check validation success, disabling alert", "crName", crName)
+		p.metricsAPI.SetUpgradeAlert(crName, 0)
```
Comment (ybettan) [inline pkg/preflight/preflight.go:290]: Isn't it the default value? do we really need to set it to zero manually?

File: pkg/preflight/preflight.go (hunk 5 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:105]: Not sure yet. Need to understand if we really need, especially since this issue won't exists in V2

File: pkg/runtime/runtime.go [hunk 22]
```diff
@@ -55,6 +55,7 @@ type RuntimeInformation struct {
 type RuntimeAPI interface {
 	GetRuntimeInformation(ctx context.Context, sr *srov1beta1.SpecialResource) (*RuntimeInformation, error)
 	LogRuntimeInformation(info *RuntimeInformation)
+	InitRuntimeInfo() *RuntimeInformation
```
Comment (ybettan) [inline pkg/runtime/runtime.go:58]: Why is that change needed?
Even if we do, I don't see any use of `InitRuntimeInfo` outside of this package - does it have to be an interface method?

File: pkg/preflight/preflight.go (hunk 5 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:105]: current code work, this might be an optimization, but it won't be relevant for V2, so i'll just remove the TODO

File: pkg/preflight/preflight.go (hunk 9 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:137]: Still thinking about it. Not sure if there is a recipe with multiple DS and BuildConfig in a chart, and it will surely not be an issue in the V2

File: pkg/preflight/preflight.go [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+
+	runInfo.OSImageURL, err = p.clusterAPI.OSImageURL(ctx)
+	if err != nil {
+		p.log.Error(err, "failed to os image url for preflight")
+		return nil, err
+	}
+
+	// no need for real proxy configuration, since preflight check do not
+	// setup any object, just verifiying images
+	runInfo.Proxy = proxy.Configuration{}
```
Comment (qbarrand) [inline pkg/preflight/preflight.go:243]: Do we need to set it then? It's not a pointer, so `p.runtimeAPI.InitRuntimeInfo` has set it to the zero value already.

File: pkg/preflight/preflight.go [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	runInfo.ClusterVersion, runInfo.ClusterVersionMajorMinor, err = p.clusterAPI.Version(ctx)
+	if err != nil {
+		p.log.Error(err, "Failed to get cluster version info for preflight")
+		return nil, err
+	}
+	// for preflight we do not need a real push secret, since preflight validation is not
+	// doing any image uploading
+	runInfo.PushSecretName = "dummyPushSecret"
```
Comment (qbarrand) [inline pkg/preflight/preflight.go:233]: Why do we need to set it?

File: pkg/preflight/preflight.go [hunk 25]
```diff
@@ -0,0 +1,295 @@
...
+	runInfo.Proxy = proxy.Configuration{}
+
+	return runInfo, nil
+}
+
+func (p *preflight) getOSData(ctx context.Context, image string) (string, string, string, error) {
+	layer, err := p.registryAPI.LastLayer(ctx, image)
+	if err != nil {
+		p.log.Error(err, "failed to get last layer of image", "image", image)
+		return "", "", "", err
```
Comment (qbarrand) [inline pkg/preflight/preflight.go:252]: WDYT about returning the error messages instead of logging them?

File: pkg/preflight/preflight.go [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+	}
+	return utils.ParseOSInfo(machineOSConfig)
+}
+
+func (p *preflight) getKernelFullVersion(ctx context.Context, image string) (string, error) {
+	layer, err := p.registryAPI.LastLayer(ctx, image)
+	if err != nil {
+		p.log.Error(err, "failed to get last layer of image", "image", image)
+		return "", err
```
Comment (qbarrand) [inline pkg/preflight/preflight.go:267]: Same remark about logging vs returning the error messages.

File: pkg/preflight/preflight.go (hunk 25 above)
Comment (ybettan) [reply pkg/preflight/preflight.go:252]: I prefer logging and returning the "virgin" error message with no wrappers around it.
Find it easier to track the root cause of an error.

When wrapping error, you may miss a few layers as it is wrapped in multiple places by multiple contributors but logging the error is easier to grep and you know the error is always the "root" error.

File: pkg/preflight/preflight.go (hunk 2 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:30]: yes, this PR just introduces a package, that will be used by the preflight controller

File: pkg/preflight/preflight.go (hunk 14 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:170]: fixed

File: pkg/preflight/preflight.go (hunk 10 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:140]: fixed

File: pkg/preflight/preflight.go (hunk 8 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:130]: fixed

File: pkg/preflight/preflight.go (hunk 7 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:125]: fixed

File: pkg/preflight/preflight.go (hunk 6 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:121]: fixed

File: pkg/preflight/preflight.go [hunk 27]
```diff
@@ -0,0 +1,295 @@
...
+
+	chart, err := p.helmerAPI.Load(sr.Spec.Chart)
+	if err != nil {
+		p.log.Info("Failed to load helm chart for CR", "name", sr.Name)
+		return err
+	}
+
+	yamlsList, err := p.processFullChartTemplates(ctx, chart, sr.Spec.Set, runInfo, sr.Namespace, runInfo.KernelFullVersion)
+	if err != nil {
+		p.log.Info("Failed to process full chart during preflight check")
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:84]: fixed

File: pkg/preflight/preflight.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:166]: fixed

File: pkg/preflight/preflight.go (hunk 1 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:73]: fixed

File: pkg/preflight/preflight.go (hunk 11 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:203]: our requirement ( will be documented in the Release docs) is that if the customer uses slim images, he must still set the etc/driver-toolkit-release.json file into the image. This can be easily done if the customer uses DTK for building (one-liner).

File: pkg/preflight/preflight.go (hunk 15 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:207]: Yes/ There are some minor fields that are not filled, since they don't really matter (like proxy, for example), because we are not actually loading the yamls into the cluster

File: pkg/preflight/preflight.go (hunk 16 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:179]: fixed

File: pkg/preflight/preflight.go (hunk 17 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:186]: fixed

File: pkg/preflight/preflight.go (hunk 18 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:191]: This is not an error from the flow point of view, we don't know in which layer DTK info is present

File: pkg/preflight/preflight.go (hunk 24 above)
Comment (ybettan) [reply pkg/preflight/preflight.go:233]: I was wondering the same thing... I can see the benefits of it but maybe it is redundant.

File: pkg/preflight/preflight.go (hunk 3 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:78]: fixed

File: pkg/preflight/preflight.go (hunk 8 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:130]: you are probably right, but i prefer to do it the same way as we do it in regular chart processing.  In any way, this code won't be needed in V2

File: pkg/preflight/preflight.go (hunk 21 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:290]: Once the CR is fixed  by the user,  SRO will pass over it again, and i case it was verified, we want to turn off the alarm

File: pkg/runtime/runtime.go (hunk 22 above)
Comment (yevgeny-shnaidman) [reply pkg/runtime/runtime.go:58]: It is used in the preflight package, to initialize the RuntimeInfo

File: pkg/preflight/preflight.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:73]: I prefer to deal with the logging in a unified epic for logging that Chris Procter will introduce

File: pkg/preflight/preflight.go [hunk 28]
```diff
@@ -0,0 +1,295 @@
...
+	chart *helmchart.Chart,
+	values unstructured.Unstructured,
+	runInfo *runtime.RuntimeInformation,
+	namespace string,
+	upgradeKernelVersion string) (string, error) {
+	var err error
+	// [TODO] - check that templates are not messed up during helm processing
+	fullChart := *chart
+	fullChart.Templates = []*helmchart.File{}
+	fullChart.Templates = append(fullChart.Templates, chart.Templates...)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:108]: i am not a fan of copy function, since it requires user to allocate the needed space for destination slice. Append takes care of that all by itself

File: pkg/preflight/preflight.go [hunk 29]
```diff
@@ -0,0 +1,295 @@
...
+//[TODO] - handle multiple daemonsets and buildconfigs
+func (p *preflight) handleYAMLsCheck(ctx context.Context, yamlsList string, upgradeKernelVersion string) (bool, error) {
+	objList, err := p.resourceAPI.GetObjectsFromYAML([]byte(yamlsList))
+	if err != nil {
+		return false, err
+	}
+	verified := true
+	for _, obj := range objList.Items {
+		kind := obj.GetKind()
+		if kind == "BuildConfig" {
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:146]: it can work but it won't be convenient: break inside switch will break only from the switch, and not from the upper loop. We can start using labels, but since it is only 2 ifs, i think staying with the ifs instead of switch should be ok

File: pkg/preflight/preflight.go [hunk 30]
```diff
@@ -0,0 +1,295 @@
...
+}
+
+func (p *preflight) daemonSetPreflightCheck(ctx context.Context, obj *unstructured.Unstructured, upgradeKernelVersion string) (bool, error) {
+	var ds k8sv1.DaemonSet
+	err := k8sruntime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &ds)
+	if err != nil {
+		p.log.Info("failed to convert YAML daemonset into sturct daemonset", "error", err)
+		return false, err
+	}
+	if ds.Spec.Template.Spec.Containers == nil || len(ds.Spec.Template.Spec.Containers) == 0 {
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:169]: fixed

File: pkg/preflight/preflight.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:179]: fixed

File: pkg/preflight/preflight.go [hunk 31]
```diff
@@ -0,0 +1,295 @@
...
+		p.log.Info("invalid daemonset, no container  data present")
+		return false, fmt.Errorf("invalid daemonset, no container  data present")
+	}
+	image := ds.Spec.Template.Spec.Containers[0].Image
+
+	p.log.Info("daemonset image for preflight validation", "image", image)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:175]: i would leave all the logger issues for Chris Procter epic

File: pkg/preflight/preflight.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:186]: fixed

File: pkg/preflight/preflight.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:191]: This is not an error, DTK info might be missing in a layer

File: pkg/preflight/preflight.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:203]: as we discussed, customer should still provide etc/driver-toolkit-release.json in the multi-stage image

File: pkg/preflight/preflight.go (hunk 23 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:243]: fixed

File: pkg/preflight/preflight.go (hunk 25 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:252]: i think we should do both, especially the first time we encounter the error

File: pkg/preflight/preflight.go (hunk 26 above)
Comment (yevgeny-shnaidman) [reply pkg/preflight/preflight.go:267]: i think we should do both, especially the first time we encounter the error

File: pkg/preflight/preflight.go (hunk 25 above)
Comment (qbarrand) [reply pkg/preflight/preflight.go:248]: I think doing both would be:
```suggestion
		p.log.Error(err, "failed to get last layer of image", "image", image)
		return "", "", "", fmt.Errorf("failed to get last layer of image %s: %v", image, err)
)
```
WDYT?

File: pkg/preflight/preflight.go (hunk 5 above)
Comment (ybettan) [reply pkg/preflight/preflight.go:104]: Are we keeping the TODO then?
